internal/policies/mount/adsys_mount/src/main.rs

```diff
@@ -0,0 +1,235 @@
+use clap::Parser;
+use gio::{
+    self,
+    traits::{FileExt, MountOperationExt},
+};
+use glib::ObjectExt;
+use log::{debug, error, warn};
+use std::{
+    fs,
+    sync::{Arc, Mutex},
+};
+
+mod logger; // Includes our logger implementation from the logger.rs file;
+use logger::Logger;
+
+mod error; // Includes our error implementation from the error.rs file;
+use error::AdsysMountError;
+
+/// Arguments required to run this binary
+#[derive(Debug, clap::Parser)]
+#[command(version, about, long_about = None)]
+struct Args {
+    /// Path for the file containing the mounts for the user.
+    mounts_file: String,
+}
+
+/// Represents a mount point read from the mounts file.
+#[derive(Debug, PartialEq, Eq)]
+struct MountEntry {
+    mount_path: String,
+    is_anonymous: bool,
+}
+
+/// Struct representing the message that is to be passed in the glib channel.
+struct Msg {
+    path: String,
+    status: MountStatus,
+}
+
+/// Represents the status returned by a mount attempt.
+#[derive(Debug)]
+enum MountStatus {
+    Done,
+    Asked,
+    Error(glib::Error),
+}
+
+fn main() -> Result<(), AdsysMountError> {
+    let args = Args::parse();
+
+    // Creates the logger and sets its level to Debug.
+    if let Ok(()) = log::set_logger(&Logger {}) {
+        log::set_max_level(log::LevelFilter::Debug);
+    }
+
+    debug!("Mounting entries listed in {}", args.mounts_file);
+
+    let parsed_entries = match parse_entries(&args.mounts_file) {
+        Ok(v) => v,
+        Err(e) => {
+            error!("Error when parsing entries: {}", e);
+            return Err(AdsysMountError::ParseError);
+        }
+    };
+
+    // Setting up the channel used for communication between the mount operations and the main function.
+    let g_ctx = glib::MainContext::default();
+    let (tx, rx) = glib::MainContext::channel(glib::PRIORITY_DEFAULT);
+
+    // Grabs the ammount of mounts to be done before passing the ownership of parsed_entries.
+    let mut mounts_left = parsed_entries.len();
+
+    for entry in parsed_entries {
+        handle_mount(entry, tx.clone());
+    }
+
+    // Sets the main loop glib to be used by the mounts
+    let g_loop = glib::MainLoop::new(Some(&g_ctx), false);
+
+    // Creates a mutex to handle the exit status
+    let mu: Arc<Mutex<Vec<Msg>>> = Arc::new(Mutex::new(Vec::new()));
+
+    // Clones the variables that are going to be moved into the closure.
+    let g_loop_clone = g_loop.clone();
+    let mu_clone = Arc::clone(&mu);
+
+    // Attaches the receiver to the main context, along with a closure that is called everytime there is a new message in the channel.
+    rx.attach(Some(&g_ctx), move |x| {
+        match x.status {
+            MountStatus::Done => debug!("Mounting of {} was successful", x.path),
+            MountStatus::Error(_) => {
+                warn!("Failed when mounting {}", x.path);
+                mu_clone.lock().unwrap().push(x);
+            }
+            _ => {}
```

Comment:
I would probably still catch the MountStatus::Ask status here to return an error that was not expected…